test/envoye2e/driver/fake_stackdriver.go

```diff
@@ -191,9 +204,111 @@ func (s *LoggingServer) GetLogEntries(w http.ResponseWriter, req *http.Request)
 	}
 }
 
+// ListTraces implements ListTraces method.
+func (s *TracesServer) ListTraces(context.Context, *cloudtracev1.ListTracesRequest) (*cloudtracev1.ListTracesResponse, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	return &s.listTracesResp, nil
+}
+
+// GetTrace implements GetTrace method.
+func (s *TracesServer) GetTrace(context.Context, *cloudtracev1.GetTraceRequest) (*cloudtracev1.Trace, error) {
+	log.Fatal("Unimplemented Method")
+	return &cloudtracev1.Trace{}, nil
+}
+
+// GetTrace implements GetTrace method.
+func (s *TracesServer) PatchTraces(context.Context, *cloudtracev1.PatchTracesRequest) (*empty.Empty, error) {
+	log.Fatal("Unimplemented Method")
+	return &empty.Empty{}, nil
+}
+
+func randomHex(n int) (string, error) {
+	bytes := make([]byte, n)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(bytes), nil
+}
+
+// BatchWriteSpans implements BatchWriteSpans method.
+func (s *TracesServer) BatchWriteSpans(ctx context.Context, req *cloudtracev2.BatchWriteSpansRequest) (*empty.Empty, error) {
+	log.Printf("receive BatchWriteSpansRequest %+v", *req)
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	var m []*cloudtracev1.Trace
+	// In first Iteration, we find all spans that are root spans and store it in an array.
+	for _, span := range req.Spans {
+		var spanID int
+		var err error
+		if spanID, err = strconv.Atoi(span.SpanId); err != nil {
+			log.Printf("err: %v", err)
+			continue
+		}
+		if span.ParentSpanId == "" {
+			traceID, _ := randomHex(32)
+			m = append(m, &cloudtracev1.Trace{
+				ProjectId: req.Name,
+				TraceId:   traceID,
```

Comment:
Trace ID should already be embedded in [span name](https://cloud.google.com/trace/docs/reference/v2/rpc/google.devtools.cloudtrace.v2#google.devtools.cloudtrace.v2.Span), no? I think it needs to use that to identify spans that belong to the same trace.

Reply:
Thanks.. I didn't know that.. this would simplify logic for identifying child spans too..